src/main/java/com/openlattice/chronicle/services/message/MessageService.kt

```diff
@@ -0,0 +1,105 @@
+package com.openlattice.chronicle.services.message
+
+import com.google.common.collect.ArrayListMultimap
+import com.google.common.collect.ImmutableSet
+import com.google.common.collect.ListMultimap
+import com.openlattice.chronicle.constants.EdmConstants
+import com.openlattice.chronicle.constants.MessageDetails
+import com.openlattice.chronicle.constants.MessageOutcome
+import com.openlattice.chronicle.services.ApiCacheManager
+import com.openlattice.chronicle.services.edm.EdmCacheManager
+import com.openlattice.chronicle.services.enrollment.EnrollmentManager
+import com.openlattice.chronicle.services.entitysets.EntitySetIdsManager
+import com.openlattice.chronicle.services.twilio.TwilioManager
+import com.openlattice.chronicle.util.ChronicleServerUtil
+import com.openlattice.client.ApiClient
+import com.openlattice.data.DataAssociation
+import com.openlattice.data.DataGraph
+import org.slf4j.LoggerFactory
+import java.time.OffsetDateTime
+import java.util.*
+import java.util.concurrent.ExecutionException
+
+/**
+ * @author toddbergman &lt;todd@openlattice.com&gt;
+ */
+class MessageService(
+    private val apiCacheManager: ApiCacheManager,
+    private val edmCacheManager: EdmCacheManager,
+    private val enrollmentManager: EnrollmentManager,
+    private val entitySetIdsManager: EntitySetIdsManager,
+    private val twilioManager: TwilioManager
+) : MessageManager {
+    override fun sendMessage(
+        organizationId: UUID,
+        studyId: UUID,
+        participantId: String,
+        messageDetails: Map<String, String>
+    ) {
+        val message = MessageDetails(
+            messageDetails?.get("type") ?: "",
+            OffsetDateTime.now(),
+            participantId,
+            messageDetails?.get("phone") ?: "",
+            messageDetails?.get("url") ?: ""
+        )
+        val outcome = twilioManager.sendMessage(participantId, message)
+        try {
+            recordMessageSent(organizationId, studyId, outcome)
+            logger.info("Recorded notification {} sent", outcome.sid)
+        } catch (e: ExecutionException) {
+            logger.error("Unable to record notification sent for SID {}", outcome.sid, e)
+        }
+    }
+
+    @Throws(ExecutionException::class)
+    fun recordMessageSent(organizationId: UUID?, studyId: UUID?, messageOutcome: MessageOutcome) {
+        val apiClient = apiCacheManager.prodApiClientCache[ApiClient::class.java]
+        val dataApi = apiClient.dataApi
+
+        // entity set ids
+        val participantES = ChronicleServerUtil.getParticipantEntitySetName(studyId)
+        val coreAppConfig = entitySetIdsManager
+            .getChronicleAppConfig(organizationId)
+        val legacyAppConfig = entitySetIdsManager
+            .getLegacyChronicleAppConfig(participantES)
+        val messageESID = coreAppConfig.messagesEntitySetId
+        val sentToESID = legacyAppConfig.sentToEntitySetId
```

Comment:
why are some of the entity sets legacy and others are not?

Reply:
Where should they be?

Reply:
I was struggling to access sentToESID from coreAppConfig

Reply:
The corresponding EntityTypeCollection doesn't have a template for the entity set. It will need to be added. CollectionsApi.addTypeToEntityTypeCollectionTemplate(). permissions config for chronicle_core app will also need to be updated: AppApi.updateAppEntitySetPermissionsConfig(). However, this call is only useful for future installations of chronicle. Eventually we will also need to update permissions for all organizations that are currently using chronicle.

Reply:
ok, so what would you like me to do right now? the code above works. only sent_to is showing in legacy, not messages. I'm not sure why, since they were both created with he legacy entity sets when setting up the app.